src/runtime/chan.go

```diff
@@ -1,520 +1,329 @@
+//   proceed.
+//   The value is at the same time also a way that goroutines can be the first
+//   (and only) goroutine to 'take' a channel operation to change it from
+//   'waiting' to any other value. This is important for the select statement
+//   because multiple goroutines could try to let different channels in the
+//   select statement proceed at the same time. By using Task.Data, only a
+//   single channel operation in the select statement can proceed.
+// - It is possible for the channel queues to contain already-processed senders
+//   or receivers. This can happen when the select statement managed to proceed
+//   but the goroutine doing the select has not yet cleaned up the stale queue
+//   entries before returning. This should therefore only happen for a short
+//   period.
 
 import (
 	"internal/task"
 	"runtime/interrupt"
 	"unsafe"
 )
 
-func chanDebug(ch *channel) {
-	if schedulerDebug {
-		if ch.bufSize > 0 {
-			println("--- channel update:", ch, ch.state.String(), ch.bufSize, ch.bufUsed)
-		} else {
-			println("--- channel update:", ch, ch.state.String())
-		}
-	}
+// The runtime implementation of the Go 'chan' type.
+type channel struct {
+	closed      bool
+	elementSize uintptr
+	bufCap      uintptr // 'cap'
+	bufLen      uintptr // 'len'
+	bufHead     uintptr
+	bufTail     uintptr
+	senders     chanQueue
+	receivers   chanQueue
+	buf         unsafe.Pointer
 }
 
-// channelBlockedList is a list of channel operations on a specific channel which are currently blocked.
-type channelBlockedList struct {
-	// next is a pointer to the next blocked channel operation on the same channel.
-	next *channelBlockedList
-
-	// t is the task associated with this channel operation.
-	// If this channel operation is not part of a select, then the pointer field of the state holds the data buffer.
-	// If this channel operation is part of a select, then the pointer field of the state holds the receive buffer.
-	// If this channel operation is a receive, then the data field should be set to zero when resuming due to channel closure.
-	t *task.Task
-
-	// s is a pointer to the channel select state corresponding to this operation.
-	// This will be nil if and only if this channel operation is not part of a select statement.
-	// If this is a send operation, then the send buffer can be found in this select state.
-	s *chanSelectState
-
-	// allSelectOps is a slice containing all of the channel operations involved with this select statement.
-	// Before resuming the task, all other channel operations on this select statement should be canceled by removing them from their corresponding lists.
-	allSelectOps []channelBlockedList
+const (
+	chanOperationWaiting = iota // waiting for a send/receive operation to continue
+	chanOperationOk             // successfully sent or received (not closed)
+	chanOperationClosed         // channel was closed, the value has been zeroed
```

Comment:
Nit: because of `chanOperationMask`, these would be clearer (to me) spelled out in binary literal form.

Reply:
Seems reasonable, thanks!